File: Codility/Python/BinaryGap.py
def solution(N):
    binary = [] # 2진법 수를 리스트로 저장
    gap = 0 # 1과 1 사이 0 갯수
    gaps = [] # gap 값을 저장하는 리스트

    # 2진법 수행
    while N > 0:
        binary.append(N % 2)
        N //= 2
    
    # binary 리스트를 뒤에서부터 탐색
    for i in reversed(binary):
        # 1이면 지금까지의 0의 갯수를 리스트에 저장 후 0으로 초기화
        if i == 1:
            gaps.append(gap)
            gap = 0
        # 0이면 0의 갯수 1씩 증가
        else:
            gap += 1
    return max(gaps)

print(solution(9)) # 2
print(solution(529)) # 4
print(solution(20)) # 1
print(solution(15)) # 0
print(solution(32)) # 0
print(solution(1041)) # 5

File: Codility/Kotlin/BinaryGap.kt
fun solution(N: Int): Int {
    var binary = N.toString(2) // 2 진법 String
    var gap: Int = 0 // 1과 1사이 0 개수
    var gaps = arrayListOf<Int>() // 1과 1사이 0 개수 리스트
    var maximum = 0 // 0 개수 최댓값

    // 2 진법 String에서 각 Char 비교
    for (i in binary) {
        // 1이 나온 경우, 0 개수 리스트에 저장 및 개수 0으로 초기화
        if (i == '1') {
            gaps.add(gap)
            gap = 0
        // 0이 나온 경우, 0 개수 1 추가
        } else {
            gap += 1
        }
    }

    // 0 개수 리스트 탐색하여 가장 큰 값 도출
    for (i in gaps) {
        if (i > maximum) {
            maximum = i
        }
    }

    return maximum
}
